zkevm-circuits/src/blake2b/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0]];
+
+fn g(v: &mut [u64; 16], a: usize, b: usize, c: usize, d: usize, x: u64, y: u64) {
+    v[a] = v[a].wrapping_add(v[b]).wrapping_add(x);
+    v[d] = (v[d] ^ v[a]).rotate_right(32);
+
+    v[c] = v[c].wrapping_add(v[d]);
+    v[b] = (v[b] ^ v[c]).rotate_right(24);
+
+    v[a] = v[a].wrapping_add(v[b]).wrapping_add(y);
+    v[d] = (v[d] ^ v[a]).rotate_right(16); 
+    
+    v[c] = v[c].wrapping_add(v[d]);
+    v[b] = (v[b] ^ v[c]).rotate_right(63);
+}
+
+fn compress(h: &mut [u64; 8], m: &[u64; 16], t: u128, f: bool, r: u32) {
+    let mut v = [0u64; 16];
+    
+    for i in 0..8 {       
+        v[i] = h[i];
+        v[i + 8] = IV[i];
+    }
+
+    v[12] ^= t as u64;
+    v[13] ^= (t >> 64) as u64;
+    
+    if f { v[14] ^= 0xFFFF_FFFF_FFFF_FFFF; }
+
+    for i in 0..(r as usize) {
+        let s = &SIGMA[i % 10];
+
+        g(&mut v, 0, 4, 8, 12, m[s[0]], m[s[1]]);
+        g(&mut v, 1, 5, 9, 13, m[s[2]], m[s[3]]);
+        g(&mut v, 2, 6, 10, 14, m[s[4]], m[s[5]]);
+        g(&mut v, 3, 7, 11, 15, m[s[6]], m[s[7]]);
+
+        g(&mut v, 0, 5, 10, 15, m[s[8]], m[s[9]]);
+        g(&mut v, 1, 6, 11, 12, m[s[10]], m[s[11]]);
+        g(&mut v, 2, 7, 8, 13, m[s[12]], m[s[13]]);
+        g(&mut v, 3, 4, 9, 14, m[s[14]], m[s[15]]);
+    }
+
+    for i in 0..8 {
+        h[i] ^= v[i] ^ v[i + 8];
+    }
+}
+
+fn load_bytes(d: &mut [u8], s: &[u64]) {
+    if s.len() * 8 < d.len() {
+        panic!("Not enough data in the source slice!");
+    }
+    
+    let q = d.len() / 8;
+    for i in 0..q {
+        d[8 * i..][..8].copy_from_slice(&s[i].to_le_bytes());
+    }
+
+    let b = d.len() - 8 * q;
+    if b == 0 { return; }
+
+    let mut l = s[q];
+    for i in 8 * q..d.len() {
+        d[i] = l as u8;
+        l >>= 8;    
+    }
+}
+
+fn save_bytes(d: &mut [u64], s: &[u8]) {
+    if d.len() * 8 < s.len() {
+        panic!("Not enough memory in the destination slice!");
+    }
+
+    let q = s.len() / 8;
+    for i in 0..q {
+        d[i] = u64::from_le_bytes(s[8 * i..][..8].try_into().unwrap());
+    }
+
+    let b = s.len() - 8 * q;
+    if b == 0 { return; }
+
+    let (mut l, mut p) = (0u64, 0);
+    for i in 8 * q..s.len() {
+        l |= (s[i] as u64) << p;
+        p += 8;
+    }
+
+    d[q] = d[q] & (0xFFFF_FFFF_FFFF_FFFF << (8 * b)) | l;
+}
+
+pub fn digest(hash :&mut [u8], msg: &[u8], key: &[u8]) {
+    if (hash.len() == 0) || (hash.len() > 64) {
+        panic!("Hash cannot be empty or longer than 64 bytes!");
+    }
+    if key.len() > 64 {
+        panic!("Key cannot be longer than 64 bytes!");
+    }
+
+    const R: u32 = 12;
+    let mut h = IV;
+
+    h[0] ^= 0x0101_0000 ^ ((key.len() as u64) << 8) ^ (hash.len() as u64);
+
+    let mut t = 0u128;
+    let s = key.len() > 0;
+    let e = msg.len() == 0;
+    let mut b = [0u64; 16];
```

Comment:
Are these names also used in the spec? If not some more descriptive ones would be nice.

Reply:
Thanks!

The last three are not in spec:
b - **b**uffer to store the data passed to the "compress" function; 
e - "**e**mpty message" flag;
s - "**s**ecret key is used" flag.
I willl add this as comments.

By the way, Github says "some checks were not successful". Is it a problem?

Reply:
Using `b` for the buffer makes sense to me because it's used a lot, but e.g. `e` is rarely used and changing it to `is_empty` would make the code more readable with no other drawbacks I think.

> By the way, Github says "some checks were not successful". Is it a problem?

I don't think so, just some tests that run on the original repo that don't work on this fork.

Reply:
Ready.